RecoMTD/DetLayers/src/MTDDetRing.cc

```diff
@@ -0,0 +1,181 @@
+  static const vector<const GeometricSearchDet*> result;
+  return result;
+}
+
+
+pair<bool, TrajectoryStateOnSurface>
+MTDDetRing::compatible(const TrajectoryStateOnSurface& ts, const Propagator& prop, 
+		      const MeasurementEstimator& est) const {
+  
+  const std::string metname = "MTD|RecoMTD|RecoMTDDetLayers|MTDDetRing";
+  TrajectoryStateOnSurface ms = prop.propagate(ts,specificSurface());
+  
+  LogTrace(metname) << "MTDDetRing::compatible, Surface at Z: " 
+                    << specificSurface().position().z()
+                    << " R1: " << specificSurface().innerRadius()
+                    << " R2: " << specificSurface().outerRadius()
+                    << " TS   at Z,R: " << ts.globalPosition().z() << ","
+                    << ts.globalPosition().perp();
+  if (ms.isValid()) {
+    LogTrace(metname) << " DEST at Z,R: " << ms.globalPosition().z() << ","
+		      << ms.globalPosition().perp()
+		      << " local Z: "   << ms.localPosition().z()  << endl;
+  }
+  else
+    LogTrace(metname) << " DEST: not valid" <<endl;
+  
+  
+  if (ms.isValid()) return make_pair(est.estimate(ms, specificSurface()) != 0, ms);
+  else return make_pair(false, ms);
+}
+
+
+vector<GeometricSearchDet::DetWithState> 
+MTDDetRing::compatibleDets( const TrajectoryStateOnSurface& startingState,
+			   const Propagator& prop, 
+			   const MeasurementEstimator& est) const {
+
+  const std::string metname = "MTD|RecoMTD|RecoMTDDetLayers|MTDDetRing";
+
+  LogTrace(metname) << "MTDDetRing::compatibleDets, Surface at Z: " 
+                    << surface().position().z()
+                    << " R1: " << specificSurface().innerRadius()
+                    << " R2: " << specificSurface().outerRadius()
+                    << " TS at Z,R: " << startingState.globalPosition().z() << ","
+                    << startingState.globalPosition().perp() << "     DetRing pos." << position();
+
+  vector<DetWithState> result;
+
+  // Propagate and check that the result is within bounds
+  pair<bool, TrajectoryStateOnSurface> compat =
+    compatible(startingState, prop, est);
+  if (!compat.first) {
+    LogTrace(metname) << "    MTDDetRing::compatibleDets: not compatible"
+                      << "    (should not have been selected!)";
+    return result;
+  }
+  
+  // Find the most probable destination component
+  TrajectoryStateOnSurface& tsos = compat.second;
+  GlobalPoint startPos = tsos.globalPosition();  
+  int closest = theBinFinder.binIndex(startPos.phi());
+  const vector<const GeomDet*> dets = basicComponents();
+  LogTrace(metname) << "     MTDDetRing::compatibleDets, closest det: " << closest 
+                    << " Phi: " << dets[closest]->surface().position().phi()
+                    << " impactPhi " << startPos.phi();
+
+  // Add this detector, if it is compatible
+  // NOTE: add performs a null propagation
+  add(closest, result, tsos, prop, est);
+
+  int nclosest = result.size(); int nnextdet=0; // MDEBUG counters
```

Comment:
nclosest comes from the add() function where the result is the list of compatible dets.
Since LogTrace zeros all the code unless EDM_ML_DEBUG is defined, I have wrapped these two variables in ifdefs.